### PR TITLE
Replace foreach with a while loop in cache expiry

### PR DESCRIPTION
**Pull request by chamil321**
Due to that exception, the clean up task crashes. So that exception has led to an OOM situation. Therefore the fix is applied to avoid the exception mentioned in the [issue](https://github.com/ballerina-platform/ballerina-lang/issues/12687)

selection_051

selection_050

### CODE REVIEW

File: stdlib/cache/src/main/ballerina/cache/cache.bal [hunk 1]
```diff
@@ -211,8 +211,18 @@ function runCacheExpiry() returns error? {
...
+    string[] currentCacheKeys = cacheMap.keys();
+    while (keyIndex < currentCacheKeys.length()) {
+
+        Cache currentCache;
+        string currentCacheKey = currentCacheKeys[keyIndex];
+        Cache? cache = cacheMap[currentCacheKey];
+        if (cache is Cache) {
+            currentCache = cache;
+        } else {
+            return ();
```
Comment (pubudu91) [inline stdlib/cache/src/main/ballerina/cache/cache.bal]: Why is there a return here? Shouldn't this be `continue`?

File: stdlib/cache/src/main/ballerina/cache/cache.bal [hunk 2]
```diff
@@ -221,8 +231,18 @@ function runCacheExpiry() returns error? {
...
+        string[] entryKeys = currentCache.entries.keys();
+        while (entrykeyIndex < entryKeys.length()) {
 
+            CacheEntry entry;
+            var key = entryKeys[entrykeyIndex];
+            CacheEntry? cacheEntry = currentCache.entries[key];
+            if (cacheEntry is CacheEntry) {
+                entry = cacheEntry;
+            } else {
+                return ();
```
Comment (pubudu91) [inline stdlib/cache/src/main/ballerina/cache/cache.bal]: Same comment as above. This should be `continue` right?

File: stdlib/cache/src/main/ballerina/cache/cache.bal [hunk 3]
```diff
@@ -231,13 +251,16 @@ function runCacheExpiry() returns error? {
...
         }
 
         // Iterate through the key list which needs to be removed.
-        foreach var currentKeyIndex in 0..<cachesToBeRemovedIndex {
+        int currentKeyIndex = 0;
+        while(currentKeyIndex < cachesToBeRemovedIndex) {
             string key = cachesToBeRemoved[currentKeyIndex];
             // Remove the cache entry.
             _ = currentCache.entries.remove(key);
+            currentKeyIndex = currentKeyIndex + 1;
```
Comment (Shan1024) [inline stdlib/cache/src/main/ballerina/cache/cache.bal]:
```suggestion
            currentKeyIndex += 1;
```

File: stdlib/cache/src/main/ballerina/cache/cache.bal [hunk 4]
```diff
@@ -246,6 +269,7 @@ function runCacheExpiry() returns error? {
             emptyCacheKeys[emptyCacheCount] = currentCacheKey;
             emptyCacheCount += 1;
         }
+        keyIndex = keyIndex + 1;
```
Comment (Shan1024) [inline stdlib/cache/src/main/ballerina/cache/cache.bal]:
```suggestion
        keyIndex += 1;
```

File: stdlib/cache/src/main/ballerina/cache/cache.bal (hunk 1 above)
Comment (chamil321) [reply stdlib/cache/src/main/ballerina/cache/cache.bal]: Yes. Resolved

**Comment by chamil321** [comment]
@pubudu91 The root cause is the ConcurrentModificationException occurs when using the iterator in enhance for loop in runCacheExpiry() method. This exception leads to crash the cacheCleanup task which eventually runs to an OOM situation.

File: stdlib/cache/src/main/ballerina/cache/cache.bal [hunk 5]
```diff
@@ -211,40 +211,60 @@ function runCacheExpiry() returns error? {
...
-            // Get the current system time.
-            int currentSystemTime = time:currentTime().time;
-
-            // Check whether the cache entry needs to be removed.
-            if (currentSystemTime >= entry.lastAccessedTime + currentCacheExpiryTime) {
-                cachesToBeRemoved[cachesToBeRemovedIndex] = key;
-                cachesToBeRemovedIndex = cachesToBeRemovedIndex + 1;
+    int keyIndex = 0;
+    string[] currentCacheKeys = cacheMap.keys();
+    while (keyIndex < currentCacheKeys.length()) {
```
Comment (wggihan) [inline stdlib/cache/src/main/ballerina/cache/cache.bal]: Shall we assign the currentCacheKeys.length() to a local variable and use that?

**Pull request merged**